snuba/datasets/discover.py

```diff
@@ -0,0 +1,278 @@
+from snuba.request.request_settings import RequestSettings
+from snuba.util import is_condition
+
+EVENTS = 'events'
+TRANSACTIONS = 'transactions'
+
+
+def detect_dataset(query: Query, transactions_columns: ColumnSet) -> str:
+    """
+    Given a query, we attempt to guess whether it is better to fetch data from the
+    "events" or "transactions" dataset. This is going to be wrong in some cases.
+    """
+    # First check for a top level condition that matches either type = transaction
+    # type != transaction.
+    conditions = query.get_conditions()
+    if conditions:
+        for idx, condition in enumerate(conditions):
+            if is_condition(condition):
+                if tuple(condition) == ('type', '!=', 'transaction'):
+                    return EVENTS
+                elif tuple(condition) == ('type', '=', 'transaction'):
+                    return TRANSACTIONS
+
+    # If there is a condition that references a transactions only field, just switch
+    # to the transactions dataset
+    if [col for col in query.get_columns_referenced_in_conditions() if transactions_columns.get(col)]:
+        return TRANSACTIONS
+
+    # Use events by default
+    return EVENTS
+
+
+class DiscoverTableSource:
+    def __init__(self):
+        self.__table_source = None
+
+    def get_table_source(self) -> Optional[TableSource]:
+        return self.__table_source
+
+    def set_table_source(self, dataset_name: str) -> None:
+        self.__table_source = get_dataset(dataset_name) \
+            .get_dataset_schemas() \
+            .get_read_schema() \
+            .get_data_source()
+
+
+class DiscoverSource(RelationalSource):
+    def __init__(self, columns: ColumnSet, discover_table_source: DiscoverTableSource) -> None:
+        self.__columns = columns
+        self.__discover_table_source = discover_table_source
+
+    def format_from(self) -> str:
+        table_source = self.__discover_table_source.get_table_source()
+        if table_source:
+            return table_source.format_from()
+        return ""
+
+    def get_columns(self) -> ColumnSet:
+        return self.__columns
+
+    def get_mandatory_conditions(self) -> Sequence[Condition]:
+        table_source = self.__discover_table_source.get_table_source()
+        if table_source:
+            return table_source.get_mandatory_conditions()
+        return []
+
+
+class DiscoverProcessor(QueryProcessor):
+    """
+    Switches the table source to either Events or Transactions, depending on
+    the detected dataset.
+    """
+
+    def __init__(self, table_source: DiscoverTableSource, transactions_columns: ColumnSet) -> None:
+        self.__table_source = table_source
+        self.__transactions_columns = transactions_columns
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        detected_dataset = detect_dataset(query, self.__transactions_columns)
+        self.__table_source.set_table_source(detected_dataset)
```

Comment:
Please di not store mutable state in the processor. There is no guarantee instances will not be reused:
https://github.com/getsentry/snuba/blob/master/snuba/query/query_processor.py#L20-L21

__transactions_columns is fine because you are not mutating that state

Why aren't you just updating the Query ?

Reply:
So I wanted this to be something like `query.get_data_source().set_table_source()` but I haven't come up with a way to do this that satisfies the type checker